// foxx-cli/dist/cmds/scripts.js
"use strict";

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const { common, serverArgs } = require("../util/cli");
const { detail, fatal, info, json } = require("../util/log");
const { ERROR_SERVICE_NOT_FOUND } = require("../errors");

const { white } = require("chalk");
const client = require("../util/client");
const { group } = require("../util/text");
const resolveServer = require("../resolveServer");

const command = exports.command = "scripts <mount>";
exports.description = "List available scripts for a mounted service";

const describe = "Fetches a list of the scripts defined by the service. Returns an object mapping the raw script names to human-friendly names.";

const args = [["mount", "Mount path of the service"]];

exports.builder = yargs => common(yargs, { command, describe, args }).options(_extends({}, serverArgs, {
  raw: {
    describe: "Output raw JSON response",
    type: "boolean",
    default: false
  }
})).example("$0 scripts /hello", 'Shows all scripts of the service at the URL "/hello"').example("$0 scripts /hello --server http://localhost:8530", "Use the server on port 8530 instead of the default").example("$0 scripts /hello --database mydb", 'Use the database "mydb" instead of the default');

exports.handler = (() => {
  var _ref = _asyncToGenerator(function* (argv) {
    try {
      const server = yield resolveServer(argv);
      const db = client(server);
      const scripts = yield db.listServiceScripts(argv.mount);
      const names = Object.keys(scripts);
      if (argv.raw) {
        json(scripts);
      } else if (names.length) {
        info(group(...names.map(function (name) {
          return [name, scripts[name]];
        })));
      } else if (argv.verbose) {
        detail("No scripts available.");
      }
    } catch (e) {
      if (e.isArangoError && e.errorNum === ERROR_SERVICE_NOT_FOUND) {
        fatal(`No service found at "${white(argv.mount)}".`);
      }
      fatal(e);
    }
  });

  function handler(_x) {
    return _ref.apply(this, arguments);
  }

  return handler;
})();
